File: translated_code/llm_client.go
// llm_client.go - 大模型 API 客户端（Go 语言版本）
package services

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"
)

// Config 定义 LLM 客户端配置
type Config struct {
	APIURL      string
	APIKey      string
	Model       string
	Temperature float64
	MaxTokens   int
	Timeout     time.Duration
}

// LLMClient 大模型客户端
type LLMClient struct {
	config Config
	client *http.Client
}

// NewLLMClient 创建 LLM 客户端实例
func NewLLMClient(config Config) *LLMClient {
	if config.Timeout == 0 {
		config.Timeout = 120 * time.Second
	}
	return &LLMClient{
		config: config,
		client: &http.Client{Timeout: config.Timeout},
	}
}

// Message 定义消息结构
type Message struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

// Request 定义请求体
type Request struct {
	Model       string    `json:"model"`
	Messages    []Message `json:"messages"`
	Temperature float64   `json:"temperature"`
	MaxTokens   int       `json:"max_tokens"`
}

// Response 定义响应体
type Response struct {
	Choices []struct {
		Message Message `json:"message"`
	} `json:"choices"`
}

// CallOption 调用选项函数类型
type CallOption func(*Config)

// WithModel 设置模型
func WithModel(model string) CallOption {
	return func(c *Config) {
		c.Model = model
	}
}

// WithTemperature 设置温度
func WithTemperature(temp float64) CallOption {
	return func(c *Config) {
		c.Temperature = temp
	}
}

// WithMaxTokens 设置最大 token 数
func WithMaxTokens(tokens int) CallOption {
	return func(c *Config) {
		c.MaxTokens = tokens
	}
}

// CallLLM 调用大模型 API
func (c *LLMClient) CallLLM(ctx context.Context, systemPrompt, userPrompt string, opts ...CallOption) (string, error) {
	// 应用可选参数
	cfg := c.config
	for _, opt := range opts {
		opt(&cfg)
	}

	// 构建请求体
	reqBody := Request{
		Model: cfg.Model,
		Messages: []Message{
			{Role: "system", Content: systemPrompt},
			{Role: "user", Content: userPrompt},
		},
		Temperature: cfg.Temperature,
		MaxTokens:   cfg.MaxTokens,
	}

	bodyBytes, err := json.Marshal(reqBody)
	if err != nil {
		return "", fmt.Errorf("序列化请求失败: %w", err)
	}

	// 创建 HTTP 请求
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, cfg.APIURL, bytes.NewReader(bodyBytes))
	if err != nil {
		return "", fmt.Errorf("创建请求失败: %w", err)
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", cfg.APIKey))

	// 发送请求
	resp, err := c.client.Do(req)
	if err != nil {
		return "", fmt.Errorf("请求失败: %w", err)
	}
	defer resp.Body.Close()

	// 检查状态码
	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return "", fmt.Errorf("API 返回错误: status=%d, body=%s", resp.StatusCode, string(body))
	}

	// 解析响应
	var result Response
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return "", fmt.Errorf("解析响应失败: %w", err)
	}

	if len(result.Choices) == 0 {
		return "", fmt.Errorf("响应中没有选择项")
	}

	return result.Choices[0].Message.Content, nil
}

// Close 关闭客户端
func (c *LLMClient) Close() {
	c.client.CloseIdleConnections()
}

File: refactored_code/llm_client.py
"""大模型 API 客户端 - 重构版"""
from __future__ import annotations

import asyncio
from typing import Protocol

import httpx

from ..config import Settings


class LLMClient(Protocol):
    """LLM 客户端协议"""
    async def call(
        self, 
        system_prompt: str, 
        user_prompt: str, 
        model: str | None = None,
        temperature: float | None = None,
        max_tokens: int | None = None,
    ) -> str: ...


class OpenAIClient:
    """OpenAI API 客户端（支持连接复用）"""

    def __init__(self, settings: Settings):
        self._settings = settings
        self._client: httpx.AsyncClient | None = None

    async def _get_client(self) -> httpx.AsyncClient:
        """获取或创建 HTTP 客户端"""
        if self._client is None or self._client.is_closed:
            self._client = httpx.AsyncClient(timeout=self._settings.llm_timeout)
        return self._client

    async def close(self) -> None:
        """关闭客户端连接"""
        if self._client and not self._client.is_closed:
            await self._client.aclose()

    async def call(
        self, 
        system_prompt: str, 
        user_prompt: str, 
        model: str | None = None,
        temperature: float | None = None,
        max_tokens: int | None = None,
    ) -> str:
        """调用大模型 API，返回模型原始响应文本"""
        model = model or self._settings.llm_model
        temperature = temperature if temperature is not None else self._settings.llm_temperature
        max_tokens = max_tokens or self._settings.llm_max_tokens

        headers = {
            "Content-Type": "application/json",
            "Authorization": f"Bearer {self._settings.get_api_key()}",
        }

        payload = {
            "model": model,
            "messages": [
                {"role": "system", "content": system_prompt},
                {"role": "user", "content": user_prompt},
            ],
            "temperature": temperature,
            "max_tokens": max_tokens,
        }

        client = await self._get_client()
        
        try:
            resp = await client.post(
                self._settings.llm_api_url, 
                json=payload, 
                headers=headers
            )
            resp.raise_for_status()
            data = resp.json()
        except httpx.HTTPStatusError as e:
            raise LLMError(f"API 请求失败: status={e.response.status_code}") from e
        except httpx.RequestError as e:
            raise LLMError(f"网络请求错误: {e}") from e
        except KeyError as e:
            raise LLMError(f"响应格式错误: 缺少字段 {e}") from e

        return data["choices"][0]["message"]["content"]


class LLMError(Exception):
    """LLM 调用错误"""
    pass


# 全局客户端实例
_client: OpenAIClient | None = None


def get_client() -> OpenAIClient:
    """获取客户端实例（单例）"""
    global _client
    if _client is None:
        from ..config import settings
        _client = OpenAIClient(settings)
    return _client
